### PR TITLE
Implement a setting service where to handle everything a little more …

### PR DESCRIPTION
**Pull request by matiasdelellis**
…cleaner.

### CODE REVIEW

File: lib/Controller/SettingController.php [hunk 1]
```diff
@@ -131,49 +132,48 @@ public function setAppValue($type, $value) {
 		$message = "";
 		switch ($type) {
 			case 'sensitivity':
-				$this->config->setAppValue('facerecognition', $type, $value);
+				$this->settingService->setSensitivity ($value);
```
Comment (stalker314314) [inline lib/Controller/SettingController.php]: extra space after `Sensitivity` :)

File: lib/Controller/SettingController.php [hunk 2]
```diff
@@ -131,49 +132,48 @@ public function setAppValue($type, $value) {
...
-						$message = $this->l10n->t("The changes were saved. It will be taken into account in the next analysis.");
-						$this->config->setAppValue('facerecognition', $type, $value);
-					}
-				} else {
+				if (!is_numeric ($value)) {
 					$status = self::STATE_ERROR;
 					$message = $this->l10n->t("The format seems to be incorrect.");
 					$value = '-1';
 				}
+				// Apply prundent limits.
```
Comment (stalker314314) [inline lib/Controller/SettingController.php]: prudent - typo

File: lib/Controller/SettingController.php [hunk 3]
```diff
@@ -131,49 +132,48 @@ public function setAppValue($type, $value) {
...
+				}
+				// If any validation error saves the value
+				if ($status !== self::STATE_ERROR) {
+					$message = $this->l10n->t("The changes were saved. It will be taken into account in the next analysis.");
+					$this->settingService->setMemoryLimits($value);
+				}
 				break;
 			case 'show-not-grouped':
-				$this->config->setAppValue('facerecognition', $type, $value);
+				$this->settingService->setShowNotGrouped($value == 'true' ? true : false);
```
Comment (stalker314314) [inline lib/Controller/SettingController.php]: `===`?

File: lib/Service/SettingService.php [hunk 4]
```diff
@@ -0,0 +1,184 @@
+<?php
```
Comment (stalker314314) [inline lib/Service/SettingService.php]: Consider naming it `SettingsService`. Not sure why, but "settinS" is usually plural in English 🤷‍♂

File: lib/Controller/SettingController.php (hunk 3 above)
Comment (matiasdelellis) [reply lib/Controller/SettingController.php]: Yes .. I still have to fine tune some things. In particular, the jsons of the controllers sometimes return boolean as text, and I would prefer that they use true booleans. This forces to do some more conversions in the code, but finally it cleans the code quite a lot.

File: lib/Service/SettingService.php (hunk 4 above)
Comment (matiasdelellis) [reply lib/Service/SettingService.php]: You're right.. :sweat_smile:

File: lib/Controller/PersonController.php [hunk 5]
```diff
@@ -22,60 +44,75 @@
...
-		$this->faceMapper   = $faceMapper;
-		$this->personMapper = $personmapper;
-		$this->userId       = $UserId;
+
+		$this->rootFolder     = $rootFolder;
+		$this->userSession    = $userSession;
+		$this->urlGenerator   = $urlGenerator;
+		$this->imageMapper    = $imageMapper;
+		$this->faceMapper     = $faceMapper;
+		$this->personMapper   = $personmapper;
```
Comment (stalker314314) [inline lib/Controller/PersonController.php]: personmapper -> personMapper?

File: lib/Controller/PersonController.php [hunk 6]
```diff
@@ -22,60 +44,75 @@
...
-		$this->urlGenerator = $urlGenerator;
-		$this->imageMapper  = $imageMapper;
-		$this->faceMapper   = $faceMapper;
-		$this->personMapper = $personmapper;
-		$this->userId       = $UserId;
+
+		$this->rootFolder     = $rootFolder;
+		$this->userSession    = $userSession;
+		$this->urlGenerator   = $urlGenerator;
+		$this->imageMapper    = $imageMapper;
```
[the remark this answers is not in the document]
Comment (matiasdelellis) [reply lib/Controller/PersonController.php]: haha.. Ok.. :wink:

File: lib/BackgroundJob/Tasks/CreateClustersTask.php [hunk 7]
```diff
@@ -208,8 +207,9 @@ private function createClusterIfNeeded(string $userId) {
 		}
 
 		// Prevents not create/recreate the clusters unnecessarily.
-		$this->config->setUserValue($userId, 'facerecognition', 'recreate-clusters', 'false');
-		$this->config->setUserValue($userId, 'facerecognition', 'force-create-clusters', 'false');
+
+		$this->settingsService->setNeedRecreateClusters(false, $userId);
+		$this->settingsService->setForceCreateClusters(false, $userId);
```
Comment (matiasdelellis) [inline lib/BackgroundJob/Tasks/CreateClustersTask.php]: @stalker314314 

I am doing a literal migration ... Line by line to not modify the logic ..
But now I observe that we have differentiated these flags, but  'force-create-clusters' never is reseted.. This is the only place where it is assigned. There is no other line that sets it to true... :confused: 

Then I try to follow the commits to see when this code was last edited.

Comment (stalker314314) [reply lib/BackgroundJob/Tasks/CreateClustersTask.php]: Oh, I used this as a helper for tests. There is comment about this:
https://github.com/matiasdelellis/facerecognition/blob/4a9e6542009d8efc9bae1a3ab0599cf59d0f3a8d/lib/BackgroundJob/Tasks/CreateClustersTask.php#L163

and it used in one test only, to force recreation:
https://github.com/matiasdelellis/facerecognition/search?q=force-create-clusters&unscoped_q=force-create-clusters

We did have it, I think, in code from users, but it was not needed at one point.

Comment (matiasdelellis) [reply lib/BackgroundJob/Tasks/CreateClustersTask.php]: Ohh.. Now remember that. Thank you!.. 😄

Comment (stalker314314) [reply lib/BackgroundJob/Tasks/CreateClustersTask.php]: But, to be honest, this solution sucks. I cannot remember why I didn't opt for some proper test mocks, but for full config value. Not sure if it was just easier/lack of time/patience, or there were some problems:/ But, looking from this distance, it is ugly:)

Comment (matiasdelellis) [reply lib/BackgroundJob/Tasks/CreateClustersTask.php]: haha.. For now do not worry .. :wink: 
Well. Talking about testing, I have to fix them again .. In this PR I broke up absolutely everyone .. :sweat_smile:

**Pull request merged**